File: src/main/java/ca/rafemurray/Course.java
package ca.rafemurray;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.block.BlockState;
import net.minecraft.registry.BuiltinRegistries;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.SimpleRegistry;
import net.minecraft.util.Identifier;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

import static ca.rafemurray.ElytraRacing.MOD_ID;
import static ca.rafemurray.ModBlockEntities.START_BLOCK_ENTITY;

/**
 * This class represents an instance of a racecourse
 */
public class Course {
    private static Map<UUID, Course> COURSES = new HashMap<>();

    private UUID id;
    private StartBlockEntity start;
    private FinishBlockEntity finish;
    private final List<BlockState> checkPoints;

    public Course() {
        this.start = null;
        this.finish = null;
        this.checkPoints = new ArrayList<>();
        this.id = UUID.randomUUID();
        COURSES.put(this.id, this);
    }

    public Course getCourse(UUID id) {
        return COURSES.get(id);
    }

    public UUID getId() {
        return id;
    }

    public String getIdString() {
        return id.toString();
    }

    public StartBlockEntity getStart(){
        return start;
    }

    public FinishBlockEntity getFinish() {
        return finish;
    }

    public void setStart(@Nullable StartBlockEntity startBlockEntity) {
        this.start = startBlockEntity;
    }

    public void setFinish(@Nullable FinishBlockEntity blockEntity) {
        this.finish = blockEntity;
    }

    // TODO: check that these are of type Checkpoint
    public void addCheckpoint(BlockState checkpoint) {
        this.checkPoints.add(checkpoint);
    }

    public boolean isValid() {
        return (finish != null && start != null);
    }

//    public static final Codec<Course> CODEC = RecordCodecBuilder.create(instance -> instance.group(
//        Codec.STRING.fieldOf("id").forGetter(Course::getIdString),
//        Registries.BLOCK_ENTITY_TYPE.getCodec().fieldOf("start").forGetter(Course::getStart)
//    ().fieldOf("start").forGetter(Course::getStart)
//    ).apply(instance, Course::new));

    @Override
    public int hashCode() {
        return id.hashCode();
    }
}

File: src/main/java/ca/rafemurray/course/Course.kt
package ca.rafemurray.course

import ca.rafemurray.ElytraRacing
import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.BlockPos
import net.minecraft.core.UUIDUtil
import net.minecraft.resources.Identifier
import net.minecraft.world.level.block.state.BlockState
import java.util.*

/**
 * This class represents an instance of a racecourse
 */
class Course // Only used for serialization. We don't need to mark as dirty since the data is coming straight from storage
private constructor(
    @JvmField val id: UUID,
    var startPos: BlockPos?,
    var finishPos: BlockPos?,
    private val checkPoints: MutableList<BlockState>
) {
    private fun setDirty() {
        CourseChangedCallback.EVENT.invoker().onCourseChanged(this)
    }

    constructor() : this(UUID.randomUUID(), null, null, ArrayList<BlockState>()) {
        setDirty()
    }

    val immutableCheckPoints: List<BlockState>
        get() = java.util.List.copyOf(checkPoints)

    fun setStart(startPos: BlockPos?) {
        this.startPos = startPos
        setDirty()
    }

    fun setFinish(finishPos: BlockPos?) {
        this.finishPos = finishPos
        setDirty()
    }

    // TODO: check that these are of type Checkpoint
    fun addCheckpoint(checkpoint: BlockState) {
        checkPoints.add(checkpoint)
        setDirty()
    }

    val isValid: Boolean
        get() = (finishPos != null && startPos != null)

    override fun hashCode(): Int {
        return id.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Course

        return id == other.id
    }

    companion object {
        @JvmField
        val COURSE_ID_IDENTIFIER: Identifier = Identifier.fromNamespaceAndPath(
            ElytraRacing.MOD_ID, "course_id"
        )
        @JvmField
        val IDENTIFIER: Identifier = Identifier.fromNamespaceAndPath(ElytraRacing.MOD_ID, "course")

        @JvmField
        val CODEC: Codec<Course> = RecordCodecBuilder.create { instance: RecordCodecBuilder.Instance<Course> ->
            instance.group(
                UUIDUtil.CODEC.fieldOf("id").forGetter { obj: Course -> obj.id },
                BlockPos.CODEC.fieldOf("start").forGetter { obj: Course -> obj.startPos },
                BlockPos.CODEC.fieldOf("finish").forGetter { obj: Course -> obj.finishPos },
                BlockState.CODEC.listOf().fieldOf("checkPoints")
                    .forGetter { obj: Course -> obj.immutableCheckPoints }
            ).apply(
                instance
            ) { id: UUID, startPos: BlockPos?, finishPos: BlockPos?, checkPoints: MutableList<BlockState> ->
                Course(
                    id,
                    startPos,
                    finishPos,
                    checkPoints
                )
            }
        }
    }
}
